nms/src/main/java/com/owen1212055/biomevisuals/nms/ApiEntityConverter.java

```diff
@@ -1,34 +1,91 @@
 package com.owen1212055.biomevisuals.nms;
 
 import com.google.gson.*;
+import com.mojang.datafixers.util.Pair;
 import com.mojang.serialization.*;
 import com.owen1212055.biomevisuals.api.types.biome.effect.*;
+import net.minecraft.core.Holder;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.sounds.SoundEvent;
 import net.minecraft.world.level.biome.*;
-import org.bukkit.craftbukkit.v1_19_R1.*;
+import org.bukkit.*;
+import org.bukkit.craftbukkit.v1_19_R2.*;
+import org.bukkit.craftbukkit.v1_19_R2.util.CraftNamespacedKey;
+
+import java.util.Objects;
 
 public class ApiEntityConverter {
 
-    public static JsonElement convert(AdditionSound sound) {
-        AmbientAdditionsSettings settings = new AmbientAdditionsSettings(CraftSound.getSoundEffect(sound.getSoundEvent()), sound.getTickChance());
+    public static JsonElement serialize(AdditionSound sound) {
+        AmbientAdditionsSettings settings = new AmbientAdditionsSettings(Holder.direct(CraftSound.getSoundEffect(sound.soundEvent())), sound.tickChance());
 
         return AmbientAdditionsSettings.CODEC.encode(settings, JsonOps.INSTANCE, JsonOps.INSTANCE.empty()).get().orThrow();
     }
 
-    public static JsonElement convert(AmbientParticle particle) {
-        AmbientParticleSettings settings = new AmbientParticleSettings(CraftParticle.toNMS(particle.getParticle(), particle.getData()), particle.getProbability());
+    public static AdditionSound deserializeAdditionSound(JsonElement sound) {
+        AmbientAdditionsSettings settings = AmbientAdditionsSettings.CODEC.decode(JsonOps.INSTANCE, sound).map(Pair::getFirst).result().orElseThrow();
+
+        return new AdditionSound(Objects.requireNonNull(nmsSoundEventToBukkit(settings.getSoundEvent().value())), settings.getTickChance());
+    }
+
+    public static JsonElement serialize(AmbientParticle particle) {
+        AmbientParticleSettings settings = new AmbientParticleSettings(CraftParticle.toNMS(particle.particle(), particle.data()), particle.probability());
 
         return AmbientParticleSettings.CODEC.encode(settings, JsonOps.INSTANCE, JsonOps.INSTANCE.empty()).get().orThrow();
     }
 
-    public static JsonElement convert(MoodSound moodSound) {
-        AmbientMoodSettings settings = new AmbientMoodSettings(CraftSound.getSoundEffect(moodSound.getSoundEvent()), moodSound.getTickDelay(), moodSound.getBlockSearchExtent(), moodSound.getSoundPositionOffset());
+    public static AmbientParticle deserializeAmbientParticle(JsonElement particle) {
+        AmbientParticleSettings settings = AmbientParticleSettings.CODEC.decode(JsonOps.INSTANCE, particle).map(Pair::getFirst).result().orElseThrow();
+
+        try {
+            // FIXME the particle data object is not serialized as such, so we can't deserialize it back
+            var probabilityField = AmbientParticleSettings.class.getDeclaredField("c");
```

Comment:
Can you put the Mojang name next to it? And var :(

Reply:
Of course, done!